File: airesources/Python3/hlt/game_map.py
from . import collision, entity


class Map:
    """
    Map which houses the current game information/metadata.
    
    :ivar my_id: Current player id associated with the map
    :ivar width: Map width
    :ivar height: Map height
    """

    def __init__(self, my_id, width, height):
        """
        :param my_id: User's id (tag)
        :param width: Map width
        :param height: Map height
        """
        self.my_id = my_id
        self.width = width
        self.height = height
        self._players = {}
        self._planets = {}

    def get_me(self):
        """
        :return: The user's player
        :rtype: Player
        """
        return self._players.get(self.my_id)

    def get_player(self, player_id):
        """
        :param int player_id: The id of the desired player
        :return: The player associated with player_id
        :rtype: Player
        """
        return self._players.get(player_id)

    def all_players(self):
        """
        :return: List of all players
        :rtype: list[Player]
        """
        return list(self._players.values())

    def get_planet(self, planet_id):
        """
        :param int planet_id:
        :return: The planet associated with planet_id
        :rtype: entity.Planet
        """
        return self._planets.get(planet_id)

    def all_planets(self):
        """
        :return: List of all planets
        :rtype: list[entity.Planet]
        """
        return list(self._planets.values())

    def nearby_entities_by_distance(self, entity):
        """
        :param entity: The source entity to find distances from
        :return: Dict containing all entities with their designated distances
        :rtype: dict
        """
        result = {}
        for foreign_entity in self._all_ships() + self.all_planets():
            if entity == foreign_entity:
                continue
            result.setdefault(entity.calculate_distance_between(foreign_entity), []).append(foreign_entity)
        return result

    def _link(self):
        """
        Updates all the entities with the correct ship and planet objects

        :return:
        """
        for celestial_object in self.all_planets() + self._all_ships():
            celestial_object._link(self._players, self._planets)

    def _parse(self, map_string):
        """
        Parse the map description from the game.

        :param map_string: The string which the Halite engine outputs
        :return: nothing
        """
        tokens = map_string.split()

        self._players, tokens = Player._parse(tokens)
        self._planets, tokens = entity.Planet._parse(tokens)

        assert(len(tokens) == 0)  # There should be no remaining tokens at this point
        self._link()

    def _all_ships(self):
        """
        Helper function to extract all ships from all players

        :return: List of ships
        :rtype: List[Ship]
        """
        all_ships = []
        for player in self.all_players():
            all_ships.extend(player.all_ships())
        return all_ships

    def _intersects_entity(self, target):
        """
        Check if the specified entity (x, y, r) intersects any planets. Entity is assumed to not be a planet.

        :param entity.Entity target: The entity to check intersections with.
        :return: The colliding entity if so, else None.
        :rtype: entity.Entity
        """
        for celestial_object in self._all_ships() + self.all_planets():
            if celestial_object is target:
                continue
            d = celestial_object.calculate_distance_between(target)
            if d <= celestial_object.radius + target.radius + 0.1:
                return celestial_object
        return None

    def obstacles_between(self, ship, target, ignore=()):
        """
        Check whether there is a straight-line path to the given point, without planetary obstacles in between.

        :param entity.Ship ship: Source entity
        :param entity.Entity target: Target entity
        :param entity.Entity ignore: Which entity type to ignore
        :return: The list of obstacles between the ship and target
        :rtype: list[entity.Entity]
        """
        obstacles = []
        entities = ([] if issubclass(entity.Planet, ignore) else self.all_planets()) \
            + ([] if issubclass(entity.Ship, ignore) else self._all_ships())
        for foreign_entity in entities:
            if foreign_entity == ship or foreign_entity == target:
                continue
            if collision.intersect_segment_circle(ship, target, foreign_entity, fudge=ship.radius + 0.1):
                obstacles.append(foreign_entity)
        return obstacles


class Player:
    """
    :ivar id: The player's unique id
    """
    def __init__(self, player_id, ships={}):
        """
        :param player_id: User's id
        :param ships: Ships user controls (optional)
        """
        self.id = player_id
        self._ships = ships

    def all_ships(self):
        """
        :return: A list of all ships which belong to the user
        :rtype: list[entity.Ship]
        """
        return list(self._ships.values())

    def get_ship(self, ship_id):
        """
        :param int ship_id: The ship id of the desired ship.
        :return: The ship designated by ship_id belonging to this user.
        :rtype: entity.Ship
        """
        return self._ships.get(ship_id)

    @staticmethod
    def _parse_single(tokens):
        """
        Parse one user given an input string from the Halite engine.

        :param list[str] tokens: The input string as a list of str from the Halite engine.
        :return: The parsed player id, player object, and remaining tokens
        :rtype: (int, Player, list[str])
        """
        player_id, *remainder = tokens
        player_id = int(player_id)
        ships, remainder = entity.Ship._parse(player_id, remainder)
        player = Player(player_id, ships)
        return player_id, player, remainder

    @staticmethod
    def _parse(tokens):
        """
        Parse an entire user input string from the Halite engine for all users.

        :param list[str] tokens: The input string as a list of str from the Halite engine.
        :return: The parsed players in the form of player dict, and remaining tokens
        :rtype: (dict, list[str])
        """
        num_players, *remainder = tokens
        num_players = int(num_players)
        players = {}

        for _ in range(num_players):
            player, players[player], remainder = Player._parse_single(remainder)

        return players, remainder

    def __str__(self):
        return "Player {} with ships {}".format(self.id, self.all_ships())

    def __repr__(self):
        return self.__str__()

File: airesources/Julia/hlt/game_map.jl
logger = get_logger("bot-logger")

struct GameMap
    id::String
    width::Int
    height::Int
    players::Dict{String, Player}
    planets::Dict{String, Planet}
end
function GameMap(id, width, height, map_string::String)
    game_map = GameMap(id, width, height, Dict{String, Player}(), Dict{String, Planet}())
    map_parse!(game_map, map_string)
    return game_map
end

"""
    map_parse!(game_map, map_string)

Parse the map description from the game. Changes definition of game_map.players and game_map.planets, so it should be used only on newly created GameMap. `map_string` contains the Halite engine outputs.
"""
function map_parse!(game_map::GameMap, map_string::String)
    tokens = Vector{String}(split(map_string))
    num_players = parse(Int, shift!(tokens))
    debug(logger, "Number of players: $num_players")
    for _ in 1:num_players
        player_id = shift!(tokens)
        player = Player(player_id)
        num_ships = parse(Int, shift!(tokens))
        for _ in 1:num_ships
            ship = Ship(player.id, tokens)
            player.ships[ship.id] = ship
            debug(logger, "Owner: $(ship.owner_id), Ship: $(ship.id), x: $(ship.x), y: $(ship.y)")
        end
        game_map.players[player_id] = player
    end
    num_planets = parse(Int, shift!(tokens))
    for _ in 1:num_planets
        planet = Planet(tokens)
        game_map.planets[planet.id] = planet
    end
    @assert isempty(tokens)
end

# Utility functions
"""
    get_player(game_map, player_id)
    get_player(game_map, entity)

Returns player defined by `player_id` or `entity.owner_id`.
"""
get_player(game_map::GameMap, player_id::String) = game_map.players[player_id]
get_player(game_map::GameMap, entity::Entity) = get_player(game_map, entity.owner_id)

"""
    get_me(game_map)

Returns the user's player.
"""
get_me(game_map::GameMap) = get_player(game_map, game_map.id)

"""
    all_players(game_map)

List of all players.
"""
all_players(game_map::GameMap) = values(game_map.players)

"""
    all_planets(game_map)

List of all planets.
"""
all_planets(game_map::GameMap) = values(game_map.planets)

"""
    get_planet(game_map, planet_id)
    get_planet(game_map, ship)

Returns planet defined by `planet_id` or `ship.docked_planet_id`.
"""
get_planet(game_map::GameMap, planet_id::String) = game_map.planets[planet_id]
get_planet(game_map::GameMap, ship::Ship) = get_planet(game_map, ship.docked_planet_id)

"""
    all_docked_ships(game_map, planet)

List of all docked ships, empty if planet is not owned.
"""
function all_docked_ships(game_map::GameMap, planet::Planet)
    planet.owned && return get_ship.(get_player(game_map, planet), planet.docked_ships_ids)
    return Vector{Ship}()
end

"""
    get_docked_ship(game_map, planet, ship_id)

Return docked ship. Throw error if ship is not docked on this planet.
"""
get_docked_ship(game_map::GameMap, planet::Planet, ship_id::String) = get_ship(get_player(game_map, planet), ship_id)

"""
    all_ships(game_map)

List of all ships in current game map.
"""
function all_ships(game_map::GameMap)
    ships = Vector{Ship}() 
    for player in all_players(game_map)
        append!(ships, all_ships(player))
    end
    return ships
end